Autosafari

Kanzen01 opened the pull request:
This script script performs actions on both safari zones :
- auto moving to closest grass tile
- auto battling (throws rock + catch if pokemon contagious and evs < 50)
- auto pick items
- auto relaunch safari
- fast animation speed during safari battles

Probably needs some more testing but it works well engouh for the moment.
Feel free to edit anything that seems wrong, since it is my first js script.

Ephenia commented:
Something that will need to be changed is the script name, it must be comprised of 3 words so they are uniform with all the other ones. Other than that briefly reviewing it looks pretty good, but it will just need testing and votes from others.

Fouppy commented on custom/autosafarizone.user.js:
```diff
@@ -0,0 +1,322 @@
+      SafariBattle.Speed.enemyTransition = 1000;
+      characterMovementSpeed = 250;
+    }
+    localStorage.setItem('safAutoPickState', fastAnimationsState);
```
```suggestion
    localStorage.setItem('fastAnimationsState', fastAnimationsState);
```

JohnyDL commented:
> auto battling (throws rock + catch if pokemon contagious and evs < 50)

>     function fightSafariPokemon() {
>         if (App.game.party.getPokemon(SafariBattle.enemy.id)?.pokerus != GameConstants.Pokerus.Uninfected && App.game.party.getPokemon(SafariBattle.enemy.id)?.evs() < 50) {
>             if (SafariBattle.enemy.angry == 0) { SafariBattle.throwRock() }
>             else { SafariBattle.throwBall(); }
>         }
>         else { SafariBattle.run(); }
>     }

is just walking back and forth in the safari zone useful? I guess it lets you hatch pokemon but very slowly....
1 I thought you only got item drops if you catch pokemon so in auto-pick items mode it might be useful to always try to catch something
2 it's always worth trying to catch pokemon you don't have
3 it's always worth trying to catch shiny pokemon
4 bait is never thrown and so those achievements are not attainable with auto safari

I propose these changes

>  function fightSafariPokemon() {
>     if (safAutoPickState == 'ON' ||
>         App.game.party.getPokemon(SafariBattle.enemy.id) == undefined ||
>         SafariBattle.enemy.id.shiny ||
>         (App.game.party.getPokemon(SafariBattle.enemy.id)?.pokerus !== GameConstants.Pokerus.Uninfected && App.game.party.getPokemon(SafariBattle.enemy.id)?.evs() < 50)) {
>       if (SafariBattle.enemy.angry === 0) { SafariBattle.throwRock(); } else { SafariBattle.throwBall(); }
>     } else {
>       if(App.game.statistics.safariBaitThrown() <= 1000 ){SafariBattle.throwBait();} else { SafariBattle.run(); }
>     }
>   }

JohnyDL commented:
Test running on chrome I've run into some bugs
Fast animations seems to break something, it simply crashes pokeclicker.... by which I mean the lights are on but eventually it doesn't do anything and refreshing the page is the only solution with pokeclicker reporting no autosave since the 'crash'

using up all the balls leads to safari restarting but a few of the elements vanish, you can have multiple animations going on the safari minigame screen and getting out of auto safari is again a reset, so you can get into the safari and then turn off the script.

![image](https://github.com/Ephenia/Pokeclicker-Scripts/assets/5202568/52aaf1fe-0199-4e4f-a0d5-31761b9b6205)
note my town map is normally below the pokemon clicking area and it's just gone, as is the safari zone screen with seeing where the player is/items/grass/automation buttons

manuww commented:
there's a more efficient way to catch pokemon on the safari by throwing a razz berry and just after a rock

K0STYURIK commented:
I tried it both on Desktop and on browser. On Desktop it prevents all other userscripts from loading, on browser there is no such conflict. But on both versions instead of throwing rock and safari ball, it just make my character walk, and then flees from every single battle. Maybe I'm just dumb and doing something wrong, but only Safari Zone achievement this script help to get is 10000 steps.

JohnyDL commented:
Yeah the browser version is borked in several ways from my experience, even with the tweak I suggested I think the problem is that it runs too fast in parts, with collecting the items it manages to collect them before the animation can walk to them and it can sometimes trigger several battles at the same time by pathing over grass while trying to get to items, if this doesn't happen the script will happily chug away with getting into battles and restarting the safari zone when it runs out of balls but if it does get into multiple battles at the same time (with either fast animations or the pathing) that's when the errors I reported before start to occur.

That said it seems like a promising start, if it can be slowed down when it's walking to do one step at a time rather than inputting all the moves in one go that would help

It completely ignores that there are some shiny pokemon sprites that it could path to which would be cool... I'd guess implementing this would be similar to items it's just checking if the sprite is a shiny pokemon rather than an item

I think I can make the change to throwing Razz berries as @manuww suggested, I could even have it throw rocks by default and throw razz berries for shinies and/or when there are more than 25 Razz berries available.... but I'll admit I'm a newb with putting things onto github and making pull requests/commits as embarrassing as that is to admit with 15 years of trying to become a coder. 

I don't know how I'd go about fixing the desktop version of the code though.

K0STYURIK commented:
Done a little bit more testing. Script stop working when item spawned in the location where it can't be picked up. Not the fault of the script probably, bad game design on the side of developers who made this possible, but I thought I should inform you anyway, in case it somehow could be prevented 
![Image 001](https://github.com/Ephenia/Pokeclicker-Scripts/assets/115414677/8474e2dd-9638-4a8e-8ef5-af65217ed859)

manuww commented:
Never seen this and I'm level 28 maybe it's only on the client version?

Le sam. 5 août 2023 à 02:49, K0STYURIK ***@***.***> a écrit :

> Done a little bit more testing. Script stop working when item spawned in
> the location where it can't be picked up. Not the fault of the script
> probably, bad game design on the side of developers who made this possible,
> but I thought I should inform you anyway, in case it somehow could be
> prevented
> [image: Image 001]
> <https://user-images.githubusercontent.com/115414677/258554321-8474e2dd-9638-4a8e-8ef5-af65217ed859.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ephenia/Pokeclicker-Scripts/pull/331#issuecomment-1666311150>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBV4XIECUJJVKNKCAPUUOD3XTWKAFANCNFSM6AAAAAA24SA5LI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

K0STYURIK commented:
> Never seen this and I'm level 28 maybe it's only on the client version? Le sam. 5 août 2023 à 02:49, K0STYURIK ***@***.***> a écrit :
> […](#)
> Done a little bit more testing. Script stop working when item spawned in the location where it can't be picked up. Not the fault of the script probably, bad game design on the side of developers who made this possible, but I thought I should inform you anyway, in case it somehow could be prevented [image: Image 001] <https://user-images.githubusercontent.com/115414677/258554321-8474e2dd-9638-4a8e-8ef5-af65217ed859.png> — Reply to this email directly, view it on GitHub <[#331 (comment)](https://github.com/Ephenia/Pokeclicker-Scripts/pull/331#issuecomment-1666311150)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BBV4XIECUJJVKNKCAPUUOD3XTWKAFANCNFSM6AAAAAA24SA5LI> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I'll test it again on browser version tomorrow when the roster of Pokemon in my Friend Safari will change (the limitations of "pokemon should be contagious and have IV<50" are really cutting down the usability of this script, as soon as the last available in the Friend Safari Pokemon reach 50 IV script is useless till next date, it is completely useless for Kanto Safari Zone), but I doubt the problem is with client version. I think the chance of getting item spawning in the blocked area of the Safari Zone is minuscule by itself and I just got something compared in rarity to full odds shiny.

JohnyDL commented:
I caught that on the browser yesterday too, oddly again level 18 but it was a different layout,but I didn't grab a screenshot

Suitov commented:
Name suggestion: Selective Safari Sifter

Behaviour suggestion: With the context that there is one mon that is only available as an item (Exeggcute Single), it would be useful to have the option of continuing to catch (because that's the only way items spawn) even after EVs reach 50.

Ideally, this might interact with Catch Filter Fantasia for full flexibility, but if _that's_ too much of a nightmare, perhaps an option to keep catching stage 3 evolutions in particular after 50 EVs? I could be wrong, but in Kalos I think there's always a stage 3 starter every day, so that would ensure the script could keep doing something (providing the player has reached the latest region) - and stage 3 evolutions tend to (from what I've seen) be best for breeding efficiency, so it's not a completely arbitrary option.

Kanzen01 commented:
Hello everyone,

I'm currently trying to take into account all your comments and I will get back to you in a few hours, and commit all these improvements.
I'm still struggling with a few features, but I'll keep digging to try and implement them.

Kanzen01 commented:
I did my best trying to take into account most of your suggestions. There are still some that I (or someone else) might take into account later.
I unfortunately won't have time to work on this script any further until next weekend. Feel free to work on it.

Here are the last updates:
- added a "throw bait" button to throw baits all the time, to get the achivement
- added a "catch all" button, to try to catch all pokemons, which permits to generate items
- pick up items (if toggled) only when only 1 ball left, and skip fights
- smoother moving animation: we now recalculate moving path after every step ; will need improvements

Unfortunately, there is something I couldn't figure why it won't work: the `checkSafariEntry` method doesn't open the new modal after a Safari reset.
I also want to completly refactor the moving method since it does not behaving as I want it to.
The last problem I had is that the pick items method now starts only when there is 1 ball left and when we are not in battle. If a fight makes us go to 1 ball and isn't finished, we need to run away if items were not picked up yet.

I didn't have enough time to test these new features in detail, but I wanted to bring them to you as soon as possible.

Suitov commented:
Looks like you're doing a fantastic job! I wouldn't know where to begin
modding a JS-based game, so colour me impressed by anything you're able to
get working for us.

-- 

On Sun, 6 Aug 2023, 19:45 Kanzen01, ***@***.***> wrote:

> I did my best trying to take into account most of your suggestions. There
> are still some that I (or someone else) might take into account later.
> I unfortunately won't have time to work on this script any further until
> next weekend. Feel free to work on it.
>
> Here are the last updates:
>
>    - added a "throw bait" button to throw baits all the time, to get the
>    achivement
>    - added a "catch all" button, to try to catch all pokemons, which
>    permits to generate items
>    - pick up items (if toggled) only when only 1 ball left, and skip
>    fights
>    - smoother moving animation: we now recalculate moving path after
>    every step ; will need improvements
>
> Unfortunately, there is something I couldn't figure why it won't work: the
> checkSafariEntry method doesn't open the new modal after a Safari reset.
> I also want to completly refactor the moving method since it does not
> behaving as I want it to.
> The last problem I had is that the pick items method now starts only when
> there is 1 ball left and when we are not in battle. If a fight makes us go
> to 1 ball and isn't finished, we need to run away if items were not picked
> up yet.
>
> I didn't have enough time to test these new features in detail, but I
> wanted to bring them to you as soon as possible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ephenia/Pokeclicker-Scripts/pull/331#issuecomment-1666942450>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMWUACS6OTGU2TKXBEOKJDXT7Q5ZANCNFSM6AAAAAA24SA5LI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

K0STYURIK commented:
Tried the new version.
- Script doesn't cause any conflict with other scripts on desktop
- Throwing bait works amazing!
- Catching everything works amazing!
- Picking items on the last ball works perfect 

the only problem is re-entering the Safari Zone when we ran out of balls, I got softlocked on both browser and desktop versions 
![Image 001](https://github.com/Ephenia/Pokeclicker-Scripts/assets/115414677/4687d66c-1cae-41d4-9953-b757dc7ed9e2)

JohnyDL commented:
I was confused with Auto pick-up items
![image](https://github.com/Ephenia/Pokeclicker-Scripts/assets/5202568/8bef446c-9ca6-4cba-86de-17eef481ad8c)
but that it happens at the end does make some sense, saves some time.

I can report Chrome with the Restarting the safari zone is still broken as with K0STYURIK the first reset it seems to continue but without the safari splash screen, the second reset I see a broken splash screen and no catch things going on

![image](https://github.com/Ephenia/Pokeclicker-Scripts/assets/5202568/a84f8c8e-049c-499b-ab56-c055e4868deb)

I'm gonna keep running to see if I can spot the other things.

Maybe what to throw should be like the pokeballs, do the selected action first and then throw safariballs if it doesn't run away
Caught -> Run
Uncaught -> Bait
Contagious -> Razz
Caught Shiny -> Rock
Uncaught Shiny -> Nanab [+ override item collection?]
Collect items -> 4 balls remaining
etc...

K0STYURIK commented:
> this might be because it closes to the map screen and it's 2 buttons to get into the safari zone not 1 and they have short animations built in, you have to click enter safari zone and then pay entrance fee... but if they're incorrectly or incompletely loaded and you click the button that will be there in half a second before it's loaded in you might get buggy behaviour.... I'm not sure but if I stop the safari on the last ball use the last ball myself and then reload with human scale pauses and clicking the buttons only once they're visible I avoid this issue.

OMG, you're right! 
I've never worked with JavaScript in my life, so I googled how to set delay in JS, applied this knowledge to AutoSafari script, and now it's successfully reload Safari Zone after running out of Safari Balls!
I've probably done it in not the most efficient way, but it works!
![Image 001](https://github.com/Ephenia/Pokeclicker-Scripts/assets/115414677/ef4193e2-a4aa-43aa-ae93-2d18d62e1aa9)

ahlfred commented:
when you are running low on balls for example 2. if you fail to catch but it doesn't flee it will throw the last ball without picking up items first

Kanzen01 commented:
Thanks for your tests, I was able to correct this first problem.
I will try to correct other ones later.

JohnyDL commented:
I'm still seeing the misplaced animations too

![image](https://github.com/Ephenia/Pokeclicker-Scripts/assets/5202568/a50f7c25-a65d-438e-a724-3854c3baa6ea)

but it does seem to be largely functional in terms of automating a lot of the Safari Zone

Intuitively I would have thought that throwing bait vs throwing rocks would have the same behaviour regarding attempts to catch but I appreciate the updates

Suitov commented:
I've been trying the script out and it works fabulously. I tried the bait feature and found it does nothing but throw bait over and over, which is hilarious. You only really need it for that one bait achievement anyhow, so no complaints here.

All else worked as expected, with items being ignored at first and picked up once we were down to one ball, which is ideal behaviour for me.

It does hang on the battle screen once in a blue moon, but it could be a ram/js issue on my oldish machine.

It's really good that the script automatically starts turned off whenever you reload the page, so that nothing happens when you aren't expecting it.

Just a note that the latest Pokeclicker update breaks it on the browser version.

K0STYURIK commented:
> Just a note that the latest Pokeclicker update breaks it on the browser version.
On desktop too.

But the latest Pokeclicker update broke so much more, SyntheticShinySynapse, IniniteSeasonalEvents, EnhancedAutoClicker and AutoSafariZone on desktop at minimum, and I haven't check every single script

JohnyDL commented:
#342 problems with the 0.10.14 it seems that to be a problem with Infinite Seasonal Events but I also encountered a seemingly separate error with Auto Safari. All the buttons appear, and it will enter the safari zone but it will not do any wandering, if I wander however it will autobattle on my behalf

Kanzen01 commented:
Just corrected bug from last update and added walking on water.

Won't have much time during next month. Feel free to update the script. 😁

Darkuwa commented:
Can we get the fast animations on the bait and rock like there is with the pokeball?

Kanzen01 commented:
@umbralOptimatum I am so sorry I did not see this PR before. It seems like I missed your other message.

Your PR has been very instructive!
Just tested it and it works like a charm.

Suitov commented:
Just noticed that the mod does attempt to catch shinies, even if they're >50 EV. Thanks, that's great.

Bad news: it looks like Magikarps (etc?) only spawn on water now, and the mod only walks on land...

Perhaps if it tried to find a grassy land tile next to a water tile and moved back and forth between the two...?! 🤷

Suitov commented:
Fantastic improvements! My little guy is now correctly grinding in water for Magikarp. So cool.

I had a question about the "Auto Seek New" toggle, which apparently seeks out uncaught pkmn. I didn't think anything uncaught would spawn inside the Safari Zone before you've obtained it elsewhere, so just out of interest, what does this setting do/change?

Suitov commented:
Oh, I completely forgot about the first Safari Zone. Never mind! Makes perfect sense.

Ponsari commented on autosafari.user.js:
The script seems just gets stuck sometimes, and I'm fairly certain that the reason is it tries to reach an item that it thinks it can't reach. The latest update made the top 3 tiles of the large trees walkable, and I don't think they are included in the "reachableGridValues" list. I also think water tiles are not included, but since items can't spawn on those it hasn't caused any problems (although an item that's surrounded by a combination of water and fences or trees could cause the same issue).

Suitov commented:
Fantastic work both. I'm trying out the new version on desktop now. With auto bait turned on, I'm already seeing a much cleaner encounter loop.

Suitov commented:
It's in the Friend Safari; I haven't checked the first Safari Zone in a
while so don't know there. It's the same animation stuff others have
mentioned.

I'll aim to get some grabs and try to note down at what point the dupes
start to appear, but that might take me a day or two - so if anyone else
gets there first or even can pinpoint the cause, you'll have my gratitude.

-- 

On Mon, 4 Sept 2023, 05:43 umbralOptimatum, ***@***.***>
wrote:

> What stuck ball animations are these? I don't think I've seen them.
> Hopefully it will be easier to address after the next main game update,
> which will hopefully include some animation speed standardization (waiting
> for them to merge my PR).
>
> I have encountered a different issue though: when the game isn't the
> active window when the final safari ball is used and the session ends, the
> game doesn't properly load the town area and hard locks. Adding a delay
> after the final ball prevented this happening while the game was focused,
> but the speeds I've tried so far apparently haven't been enough at
> background processing speed. Can anyone else reproduce this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ephenia/Pokeclicker-Scripts/pull/331#issuecomment-1704601752>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMWUAEBUWYKFVCX2RHPRV3XYVMADANCNFSM6AAAAAA24SA5LI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

CyrilG93 commented:
I got the two issue.
Animation of ball  rotating stacking up on different place (even after the end of a safari), and the hard stuck at the end.
But the lock only happend if Auto Fast Anim is enable.
And the ball animation happend with and without it.
(see image https://imgur.com/Jhis2AI not the worse one, sometimes, theres like 3-4 spot with ball rotating on it)
(Opera GX browser btw)

But anyway, it work for what i need, great work !

Ephenia commented:
We'll merge this now!
